server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -2919,8 +2920,10 @@ private void doSnapshotShard(SnapshotShardContext context) {
                 };
             }
 
+            // filesToSnapshot will be emptied while snapshotting the file. We make a copy here for cleanup purpose in case of failure.
+            final AtomicReference<List<FileInfo>> fileToCleanUp = new AtomicReference<>(List.copyOf(filesToSnapshot));
             final ListenableFuture<Collection<Void>> allFilesUploadedListener = new ListenableFuture<>();
-            allFilesUploadedListener.addListener(context.delegateFailureAndWrap((delegate, v) -> {
+            allFilesUploadedListener.addListener(ActionListener.wrap(ignore -> {
```

Comment:
somewhat-unrelated nit: this is equivalent (and simpler) right?

```
            final ActionListener<Collection<Void>> allFilesUploadedListener = ActionListener.assertOnce(ActionListener.wrap(ignore -> {
```

Reply:
I was concerned that changing it from `ListenableFuture` to a plain `ActionListener` could cause test failures due to how exception is wrapped by `ListenableFuture`. But running some tests locally seems to be fine. I have updated the PR with the suggestion and will see whether CI is happy with it.